toolkit/tools/internal/timestamp/timestamp_mgr.go

```diff
@@ -0,0 +1,365 @@
+	mgr.EventQueue <- &TimeStampRecord{EventType: eventType, TimeStamp: ts, time: time.Now()}
+}
+
+// Process each event submitted to the queue by updating in-memory data structure and recording to file
+func (mgr *TimeStampManager) processEventsInQueue() {
+	for event := range mgr.EventQueue {
+		mgr.updateRead(event)
+		mgr.writeToFile(event)
+	}
+	mgr.eventProcessorFinished <- true
+}
+
+// Append a timestamp record to file, and periodically flush to disk
+func (writeMgr *TimeStampWriteManager) writeToFile(record *TimeStampRecord) {
+	outputBytes, err := json.Marshal(record)
+	if err != nil {
+		logger.Log.Warnf("Failed to marshal timestamp record: %v", err)
+		return
+	}
+	writeMgr.writeBuffer = append(writeMgr.writeBuffer, outputBytes)
+
+	cooldownDuration := time.Duration(writeCooldownMilliseconds) * time.Millisecond
+	if writeMgr.lastWrite.Add(cooldownDuration).After(time.Now()) {
+		return
+	}
+
+	writeMgr.flush()
+	writeMgr.lastWrite = time.Now()
+}
+
+// Flush write buffer to disk
+func (writeMgr *TimeStampWriteManager) flush() {
+	for _, outputBytes := range writeMgr.writeBuffer {
+		_, err := writeMgr.fileDescriptor.WriteString(string(outputBytes) + "\n")
+		if err != nil {
+			logger.Log.Warnf("Failed to write timestamp record to file %s: %v", writeMgr.filePath, err)
+			continue
+		}
+	}
+	err := writeMgr.fileDescriptor.Sync()
+	if err != nil {
+		logger.Log.Warnf("Failed to save file %s to disk: %v", writeMgr.filePath, err)
+		return
+	}
+	writeMgr.writeBuffer = nil
+}
+
+// Update the timestamp tree for each event submitted to the event queue
+func (mgr *TimeStampManager) updateRead(record *TimeStampRecord) {
+	if record.TimeStamp == nil {
+		record.TimeStamp = mgr.lastVisited
+	}
+
+	switch record.EventType {
+	case EventStart:
+		ts := record.TimeStamp
+		ts.StartTime = &record.time
+		mgr.nodes[ts.ID] = ts
+		if record.parentTimestamp == nil {
+			if mgr.lastVisited == nil {
+				mgr.root = ts
+			} else {
+				mgr.lastVisited.addSubStep(ts)
+			}
+			mgr.lastVisited = ts
+		} else {
+			mgr.nodes[ts.ParentID].addSubStep(ts)
+			if mgr.lastVisited.ID == ts.ParentID {
+				mgr.lastVisited = ts
+			}
+		}
+	case EventStop:
+		ts := mgr.nodes[record.ID]
+		ts.complete(record.time)
+		if record.EventType == EventStop && ts.ID == mgr.lastVisited.ID {
+			mgr.lastVisited = ts.parentTimestamp
+		}
+	}
+	return
+}
+
+// Read records written to a file and build a (partially) finished timestamp tree. This is useful for resuming
+// partial recording progress
+func (readMgr *TimeStampReadManager) buildTreeFromFile(fd *os.File) (err error) {
+	fd.Seek(0, 0)
+	scanner := bufio.NewScanner(fd)
+	for scanner.Scan() {
+		var ts TimeStamp
+		err = json.Unmarshal(scanner.Bytes(), &ts)
+		if err != nil {
+			logger.Log.Warnf("Error reading timestamp object from file %v", err)
+			continue
+		}
+		readMgr.nodes[ts.ID] = &ts
+		if readMgr.root == nil {
+			readMgr.root = &ts
+		} else {
+			readMgr.nodes[ts.ParentID].addSubStep(&ts)
+		}
```

Comment:
Ran into some nil ptr exceptions while tinkering with this. I think this works as expected now?
```suggestion
		} else if ts.ParentID != -1 {
			readMgr.nodes[ts.ParentID].addSubStep(&ts)
			readMgr.lastVisited = &ts
		}
```

Reply:
Do you have steps to replicate this error? With our current way of using timestamp manager I don't think we ever have more than one node without a parent ID

Reply:
I created some nodes, saved them to a file, then  reloaded them from file and continued. I don't think I've got the code anymore though, was just a quick hack in a unit test.